File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int left = 0, right = n-1, left_max = 0, right_max = 0;
        int rain = 0;
        while(left <= right) {
            if(left_max <= right_max) {
                if(height[left] > left_max) left_max = height[left];
                rain += left_max - height[left++]; 
            }
            else {
                if(height[right] > right_max) right_max = height[right];
                rain += right_max - height[right--];
            }
        }
        return rain;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] arr) {
        int n = arr.length;
        int rain = 0;
        int left = 0, right = n-1;
        int left_max = 0, right_max = 0;
        while(left <= right) {
            if(left_max <= right_max) {
                left_max = Math.max(left_max,arr[left]);
                rain += left_max - arr[left];
                left++;
            }
            else {
                right_max = Math.max(right_max,arr[right]);
                rain += right_max - arr[right];
                right--;
            }
        }
        return rain;
    }
}